src/service_models.rs

```diff
@@ -0,0 +1,749 @@
+    pub ports: Option<Vec<EndpointPortConfig<T>>>,
+}
+
+/// The mode of resolution to use for internal load balancing between tasks.
+#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+#[allow(missing_docs)]
+pub enum EndpointSpecMode {
+    Vip,
+    Dnsrr,
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+#[allow(missing_docs)]
+pub struct EndpointPortConfig<T>
+where
+    T: AsRef<str> + Eq + Hash,
+{
+    pub name: Option<T>,
+    pub protocol: Option<EndpointPortConfigProtocol>,
+    /// The port inside the container.
+    pub target_port: Option<isize>,
+    /// The port on the swarm hosts.
+    pub published_port: Option<isize>,
+    /// The mode in which port is published.  <p><br /></p>  - \"ingress\" makes the target port accessible on every node,   regardless of whether there is a task for the service running on   that node or not. - \"host\" bypasses the routing mesh and publish the port directly on   the swarm node where that service is running.
+    pub publish_mode: Option<EndpointPortConfigPublishMode>,
+}
+
+#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+#[allow(missing_docs)]
+pub enum EndpointPortConfigProtocol {
+    Tcp,
+    Udp,
+    Sctp,
+}
+
+/// The mode in which port is published.  <p><br /></p>  - \"ingress\" makes the target port accessible on every node,   regardless of whether there is a task for the service running on   that node or not. - \"host\" bypasses the routing mesh and publish the port directly on   the swarm node where that service is running.
+#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+#[allow(missing_docs)]
+pub enum EndpointPortConfigPublishMode {
+    Ingress,
+    Host,
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+#[allow(missing_docs)]
+pub struct ServiceEndpoint<T>
+where
+    T: AsRef<str> + Eq + Hash,
+{
+    pub spec: Option<EndpointSpec<T>>,
+    pub ports: Option<Vec<EndpointPortConfig<T>>>,
+    #[serde(rename = "VirtualIPs")]
+    pub virtual_ips: Option<Vec<ServiceEndpointVirtualIPs<T>>>,
+}
+
+#[derive(Debug, Clone, Default, Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+#[allow(missing_docs)]
+pub struct ServiceEndpointVirtualIPs<T>
+where
+    T: AsRef<str> + Eq + Hash,
+{
+    #[serde(rename = "NetworkID")]
+    pub network_id: Option<T>,
+    pub addr: Option<T>,
+}
+
+/// The status of a service update.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+#[allow(missing_docs)]
+pub struct ServiceUpdateStatus<T>
+where
+    T: AsRef<str> + Eq + Hash,
+{
+    pub state: ServiceUpdateStatusState,
+    #[serde(with = "ts_seconds")]
+    pub started_at: DateTime<Utc>,
+    #[serde(with = "ts_seconds")]
+    pub completed_at: DateTime<Utc>,
```

Comment:
The two `ts_seconds` annotation breaks JSON parsing in service listing for me. Removing them allows me to make successful `list_services` calls. On my Docker, these have nanoseconds.
```
Client:
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.4
 Git commit:        2d0083d
 Built:             Fri Aug 16 14:19:32 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.7
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.4
  Git commit:       2d0083d
  Built:            Wed Aug 14 19:38:45 2019
  OS/Arch:          linux/amd64
  Experimental:     false

```

Reply:
Thanks for this. This PR needs some integration tests, before we can merge it into master.

Reply:
@bittrance thanks for trying this out already. i've seen you also found a few other problems, is it okay for you if i pull your changes into this pr?